src/gui/qgsquerybuilder.cpp

```diff
@@ -436,3 +452,136 @@ void QgsQueryBuilder::showHelp()
 {
   QgsHelp::openHelp( QStringLiteral( "working_with_vector/vector_properties.html#query-builder" ) );
 }
+
+void QgsQueryBuilder::saveQuery()
+{
+  QgsSettings s;
+  QString lastQueryFileDir = s.value( QStringLiteral( "/UI/lastQueryFileDir" ), QDir::homePath() ).toString();
+  //save as qqf (QGIS query file)
+  QString saveFileName = QFileDialog::getSaveFileName( nullptr, tr( "Save Query to File" ), lastQueryFileDir, tr( "Query files (*.qqf *.QQF)" ) );
+  if ( saveFileName.isNull() )
+  {
+    return;
+  }
+
+  if ( !saveFileName.endsWith( QLatin1String( ".qqf" ), Qt::CaseInsensitive ) )
+  {
+    saveFileName += QLatin1String( ".qqf" );
+  }
+
+  QFile saveFile( saveFileName );
+  if ( !saveFile.open( QIODevice::WriteOnly | QIODevice::Truncate ) )
+  {
+    QMessageBox::critical( nullptr, tr( "Save Query to File" ), tr( "Could not open file for writing." ) );
+    return;
+  }
+
+  QDomDocument xmlDoc;
+  QDomElement queryElem = xmlDoc.createElement( QStringLiteral( "Query" ) );
+  QDomText queryTextNode = xmlDoc.createTextNode( txtSQL->text() );
+  queryElem.appendChild( queryTextNode );
+  xmlDoc.appendChild( queryElem );
+
+  QTextStream fileStream( &saveFile );
+  xmlDoc.save( fileStream, 2 );
+
+  QFileInfo fi( saveFile );
+  s.setValue( QStringLiteral( "/UI/lastQueryFileDir" ), fi.absolutePath() );
+}
+
+void QgsQueryBuilder::loadQuery()
+{
+  QgsSettings s;
+  QString lastQueryFileDir = s.value( QStringLiteral( "/UI/lastQueryFileDir" ), QDir::homePath() ).toString();
+
+  QString queryFileName = QFileDialog::getOpenFileName( nullptr, tr( "Load Query from File" ), lastQueryFileDir, tr( "Query files" ) + " (*.qqf);;" + tr( "All files" ) + " (*)" );
+  if ( queryFileName.isNull() )
+  {
+    return;
+  }
+
+  QFile queryFile( queryFileName );
+  if ( !queryFile.open( QIODevice::ReadOnly ) )
+  {
+    QMessageBox::critical( nullptr, tr( "Load Query from File" ), tr( "Could not open file for reading." ) );
+    return;
+  }
+  QDomDocument queryDoc;
+  if ( !queryDoc.setContent( &queryFile ) )
+  {
+    QMessageBox::critical( nullptr, tr( "Load Query from File" ), tr( "File is not a valid xml document." ) );
+    return;
+  }
+
+  QDomElement queryElem = queryDoc.firstChildElement( QStringLiteral( "Query" ) );
+  if ( queryElem.isNull() )
+  {
+    QMessageBox::critical( nullptr, tr( "Load Query from File" ), tr( "File is not a valid query document." ) );
+    return;
+  }
+
+  QString query = queryElem.text();
+
+  //TODO: test if all the attributes are valid
+  QgsExpression search( query );
+  if ( search.hasParserError() )
+  {
+    QMessageBox::critical( this, tr( "Query Result" ), search.parserErrorString() );
+    return;
+  }
+
+  QString newQueryText = query;
+
+#if 0
+  // TODO: implement with visitor pattern in QgsExpression
```

Comment:
What's the purpose of this section exactly?

Reply:
This potentially can be useful for cases when loaded query reference attributes which are not exists in the layer. In such case we may want to try "fix" query by asking user which attributes should be used instead of missed ones and making corresponding replacement.

On other hand we don't do this in other places, so maybe it should be removed.